frmts/vrt/gdaltileindexdataset.cpp

```diff
@@ -0,0 +1,3642 @@
+
+        double dfXOff = (sEnvelope.MinX - m_adfGeoTransform[GT_TOPLEFT_X]) /
+                        m_adfGeoTransform[GT_WE_RES];
+        if (dfXOff >= nRasterXSize)
+            continue;
+
+        double dfYOff = (sEnvelope.MaxY - m_adfGeoTransform[GT_TOPLEFT_Y]) /
+                        m_adfGeoTransform[GT_NS_RES];
+        if (dfYOff >= nRasterYSize)
+            continue;
+
+        double dfXSize =
+            (sEnvelope.MaxX - sEnvelope.MinX) / m_adfGeoTransform[GT_WE_RES];
+        if (dfXOff < 0)
+        {
+            dfXSize += dfXOff;
+            dfXOff = 0;
+            if (dfXSize <= 0)
+                continue;
+        }
+
+        double dfYSize = (sEnvelope.MaxY - sEnvelope.MinY) /
+                         std::fabs(m_adfGeoTransform[GT_NS_RES]);
+        if (dfYOff < 0)
+        {
+            dfYSize += dfYOff;
+            dfYOff = 0;
+            if (dfYSize <= 0)
+                continue;
+        }
+
+        const char *pszTileName =
+            poFeature->GetFieldAsString(m_nLocationFieldIndex);
+        const std::string osTileName(
+            GetAbsoluteFileName(pszTileName, GetDescription()));
+        VSIStatBufL sStatSource;
+        if (VSIStatL(osTileName.c_str(), &sStatSource) != 0 ||
+            sStatSource.st_mtime <= mTime)
+        {
+            continue;
+        }
+
+        constexpr double EPS = 1e-8;
+        GTISourceDesc oSourceDesc;
+        oSourceDesc.osFilename = osTileName;
+        oSourceDesc.nDstXOff = static_cast<int>(dfXOff + EPS);
+        oSourceDesc.nDstYOff = static_cast<int>(dfYOff + EPS);
+        oSourceDesc.nDstXSize = static_cast<int>(dfXSize + 0.5);
+        oSourceDesc.nDstYSize = static_cast<int>(dfYSize + 0.5);
+        oRes.emplace_back(std::move(oSourceDesc));
+    }
+
+    return oRes;
+}
+
+/************************************************************************/
+/*                         GetSourceDesc()                              */
+/************************************************************************/
+
+bool GDALTileIndexDataset::GetSourceDesc(const std::string &osTileName,
+                                         SourceDesc &oSourceDesc)
+{
+    std::shared_ptr<GDALDataset> poTileDS;
+    if (!m_oMapSharedSources.tryGet(osTileName, poTileDS))
+    {
+        struct Releaser
+        {
+            void operator()(GDALDataset *poDS)
+            {
+                if (poDS)
+                    poDS->Release();
+            }
+        };
+
+        poTileDS = std::shared_ptr<GDALDataset>(
+            GDALProxyPoolDataset::Create(
+                osTileName.c_str(), nullptr, GA_ReadOnly,
+                /* bShared = */ true, m_osUniqueHandle.c_str()),
+            Releaser());
+        if (!poTileDS || poTileDS->GetRasterCount() == 0)
+        {
+            return false;
+        }
+
+        // do palette -> RGB(A) expansion
+        if (poTileDS->GetRasterCount() == 1 && (nBands == 3 || nBands == 4) &&
```

Comment:
Code looks very similar to https://github.com/OSGeo/gdal/pull/8983/files#diff-40c5a795fa9c026c21ba307c0ec08fc6bc1d0bc18532dba9ac1e3be2e21f4389R871 is there a chance to factor this out in a separate method?

Reply:
done. Factored in a GTIDoPaletteExpansionIfNeeded() method